src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -26,10 +27,21 @@ public IEnumerator<TElement> GetEnumerator()
             Buffer<TElement> buffer = new Buffer<TElement>(_source);
             if (buffer._count > 0)
             {
-                int[] map = SortedMap(buffer);
-                for (int i = 0; i < buffer._count; i++)
+                int[] map = null;
+                try
                 {
-                    yield return buffer._items[map[i]];
+                    map = SortedMap(buffer);
+                    for (int i = 0; i < buffer._count; i++)
+                    {
+                        yield return buffer._items[map[i]];
+                    }
+                }
+                finally
```

Comment:
> do we strictly need to guarantee the return of the rented array no matter what? 

Not strictly; and this currently wouldn't guarantee that e.g. if someone used the iterator out of the common constructs and didn't dispose it.

> Can we just return without putting this into a finally?

This would make not returning a what's likely a Gen2 buffer more of a common case. However, is code with a broken comparer/allocator such a major even otherwise that not returning a buffer pretty minor by comparison to the overall exception that will be happening?